web_external/IndexView/index.jsx

```diff
@@ -1,21 +1,100 @@
 import React, { Component } from 'react';
+import events from 'girder/events';
+
 import TreeView from '../TreeView';
 import TrackAttribute from '../TrackAttribute';
+import annotationGeometryParser from '../util/annotationGeometryParser';
+import annotationActivityParser from '../util/annotationActivityParser';
+import annotationTrackParser from '../util/annotationTrackParser';
+import { restRequest } from 'girder/rest';
 import Viewer from '../Viewer';
 
 import './style.styl';
 
 class IndexView extends Component {
     constructor(props) {
         super(props);
+        this.toggleActivity = this.toggleActivity.bind(this);
+        this.toggleTrack = this.toggleTrack.bind(this);
+        this.state = {
+            annotationActivityContainer: null,
+            annotationTrackContainer: null,
+            annotationGeometryContainer: null
+        }
+    }
+    componentDidMount() {
+        // the states would be lift to redux store
+        events.on('v:item_selected', (itemModel) => {
+            this.setState({ itemModel });
+            downloadItemByName(itemModel.get('folderId'), 'activities.yml')
+                .then((raw) => {
+                    var annotationActivityContainer = annotationActivityParser(raw);
+                    this.setState({ annotationActivityContainer });
+                });
+            downloadItemByName(itemModel.get('folderId'), 'types.yml')
+                .then((raw) => {
+                    var annotationTrackContainer = annotationTrackParser(raw);
+                    this.setState({ annotationTrackContainer });
+                });
+            downloadItemByName(itemModel.get('folderId'), 'geom.yml')
+                .then((raw) => {
+                    var annotationGeometryContainer = annotationGeometryParser(raw);
+                    this.setState({ annotationGeometryContainer });
+                }).catch(() => {
+                    this.setState({ annotationGeometryContainer: false })
+                    events.trigger('g:alert', {
+                        icon: 'ok',
+                        text: 'Didn\'t find annotation files',
+                        type: 'danger',
+                        timeout: 4000
+                    });
+                });
+        });
     }
+
+    toggleActivity(activity, enabled) {
+        var annotationActivityContainer = this.state.annotationActivityContainer;
+        var annotationActivityContainer = annotationActivityContainer.toggleState(activity.id2, enabled);
+        this.setState({ annotationActivityContainer });
```

Comment:
As a note, this seems like a massively expensive way to effect a minor change?

Reply:
Which action do you think is expensive?

Reply:
Actually... what does this even *do*? It looks like it appends the key/values of the container to `this.state`...?

What it *shouldn't* do is trigger `componentWillReceiveProps` on the container, which is what I thought it was doing when I first looked. Now, I'm wondering if the last line just doesn't belong...

Reply:
I think it's just how React work.

Reply:
Yeah... apparently JSX treats an identifier by itself inside `{}`s as a key/value pair with key == the identifier name, value == the identifier value. Confusingly, straight Javascript treats `{foo}` as being the same as `foo`...